src/components/sidenav/sidenav.ts

```diff
@@ -0,0 +1,353 @@
+  @Output('close-start') onCloseStart = new EventEmitter<void>();
+
+  /** Event emitted when the sidenav is fully closed. */
+  @Output('close') onClose = new EventEmitter<void>();
+
+
+  /**
+   * Constructor.
+   * @param elementRef_ The DOM element reference. If not available we do not hook on transitions.
+   */
+  constructor(private elementRef_: ElementRef) {}
+
+  /**
+   * Whether the sidenav is opened. We overload this because we trigger an event when it
+   * starts or end.
+   * @returns {boolean}
+   */
+  get opened(): boolean { return this.opened_; }
+  set opened(v: boolean) {
+    this.toggle(v);
+  }
+
+
+  /** Open this sidenav, and return a Promise that will resolve when it's fully opened (or get
+   * rejected if it didn't). */
+  open(): Promise<void> {
+    return this.toggle(true);
+  }
+
+  /**
+   * Close this sidenav, and return a Promise that will resolve when it's fully closed (or get
+   * rejected if it didn't).
+   */
+  close(): Promise<void> {
+    return this.toggle(false);
+  }
+
+  /**
+   * Toggle this sidenav. This is equivalent to calling open() when it's already opened, or
+   * close() when it's closed.
+   * @param isOpen
+   * @returns {Promise<void>}
+   */
+  toggle(isOpen: boolean = !this.opened): Promise<void> {
+    // Shortcut it if we're already opened.
+    if (isOpen === this.opened) {
+      if (!this.transition_) {
+        return Promise.resolve();
+      } else {
+        return isOpen ? this.openPromise_ : this.closePromise_;
+      }
+    }
+
+    this.opened_ = isOpen;
+    this.transition_ = true;
+
+    if (isOpen) {
+      this.onOpenStart.emit(null);
+    } else {
+      this.onCloseStart.emit(null);
+    }
+
+    const emitter = isOpen ? this.onOpen : this.onClose;
+    const other = isOpen ? this.onClose : this.onOpen;
+
+    if (isOpen) {
+      if (!this.openPromise_) {
+        this.openPromise_ = new Promise<void>((resolve, reject) => {
+          this.openPromiseResolve_ = resolve;
+          this.openPromiseReject_ = reject;
+        });
+      }
+      return this.openPromise_;
+    } else {
+      if (!this.closePromise_) {
+        this.closePromise_ = new Promise<void>((resolve, reject) => {
+          this.closePromiseResolve_ = resolve;
+          this.closePromiseReject_ = reject;
+        });
+      }
+      return this.closePromise_;
+    }
+  }
+
+
+  /**
+   * When transition has finished, set the internal state for classes and emit the proper event.
+   * The event passed is actually of type TransitionEvent, but that type is not available in
+   * Android so we use any.
+   * @param e The event.
+   * @private
+   */
+  @HostListener('transitionend', ['$event']) private onTransitionEnd_(e: any) {
+    if (e.target == this.elementRef_.nativeElement
+        // We do use the webkit-transform style, so some browsers will use this as the propertyName.
+        && (e.propertyName == 'transform' || e.propertyName == '-webkit-transform')) {
```

Comment:
You could use a regex for `transform`
(I checked, ts2dart supports regex literals)